File: crates/geodb-ffi/geodb_flutter/macos/Classes/GeodbFlutterPlugin.swift
import Cocoa
import FlutterMacOS
import GeodbKit  // Import the SPM package

public class GeodbFlutterPlugin: NSObject, FlutterPlugin {
    private var geodb: GeoDbEngine?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "geodb_flutter", binaryMessenger: registrar.messenger)
        let instance = GeodbFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            handleInitialize(result: result)

        case "getStats":
            handleGetStats(result: result)

        case "getCountryCount":
            handleGetCountryCount(result: result)

        case "findCountryByCode":
            guard let args = call.arguments as? [String: Any],
                  let code = args["code"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing 'code'", details: nil))
                return
            }
            handleFindCountryByCode(code: code, result: result)

        case "findCountriesBySubstring":
            guard let args = call.arguments as? [String: Any],
                  let substring = args["substring"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing 'substring'", details: nil))
                return
            }
            handleFindCountriesBySubstring(substring: substring, result: result)

        case "findStatesBySubstring":
            guard let args = call.arguments as? [String: Any],
                  let substring = args["substring"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing 'substring'", details: nil))
                return
            }
            handleFindStatesBySubstring(substring: substring, result: result)

        case "findCitiesBySubstring":
            guard let args = call.arguments as? [String: Any],
                  let substring = args["substring"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing 'substring'", details: nil))
                return
            }
            handleFindCitiesBySubstring(substring: substring, result: result)

        case "findNearest":
            guard let args = call.arguments as? [String: Any],
                  let lat = args["lat"] as? Double,
                  let lng = args["lng"] as? Double,
                  let count = args["count"] as? Int else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing lat, lng, or count", details: nil))
                return
            }
            handleFindNearest(lat: lat, lng: lng, count: count, result: result)

        case "findInRadius":
            guard let args = call.arguments as? [String: Any],
                  let lat = args["lat"] as? Double,
                  let lng = args["lng"] as? Double,
                  let radiusKm = args["radiusKm"] as? Double else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing lat, lng, or radiusKm", details: nil))
                return
            }
            handleFindInRadius(lat: lat, lng: lng, radiusKm: radiusKm, result: result)

        case "smartSearch":
            guard let args = call.arguments as? [String: Any],
                  let query = args["query"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing 'query'", details: nil))
                return
            }
            handleSmartSearch(query: query, result: result)

        case "getPlatformVersion":
            result("macOS " + ProcessInfo.processInfo.operatingSystemVersionString)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Handler Methods

    private func handleInitialize(result: @escaping FlutterResult) {
        do {
            geodb = try GeoDbEngine()
            result(nil)
        } catch {
            result(FlutterError(code: "INIT_ERROR",
                              message: "Failed to initialize GeoDB",
                              details: error.localizedDescription))
        }
    }

    private func handleGetStats(result: @escaping FlutterResult) {
        guard let geodb = geodb else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "GeoDB not initialized", details: nil))
            return
        }

        let stats = geodb.stats()
        result([
            "countries": stats.countries,
            "states": stats.states,
            "cities": stats.cities
        ])
    }

    private func handleGetCountryCount(result: @escaping FlutterResult) {
        guard let geodb = geodb else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "GeoDB not initialized", details: nil))
            return
        }

        result(Int(geodb.countryCount()))
    }

    private func handleFindCountryByCode(code: String, result: @escaping FlutterResult) {
        guard let geodb = geodb else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "GeoDB not initialized", details: nil))
            return
        }

        if let country = geodb.findCountryByCode(code: code) {
            result(cityResultToMap(country))
        } else {
            result(nil)
        }
    }

    private func handleFindCountriesBySubstring(substring: String, result: @escaping FlutterResult) {
        guard let geodb = geodb else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "GeoDB not initialized", details: nil))
            return
        }

        let countries = geodb.findCountriesBySubstring(substr: substring)
        result(countries.map { cityResultToMap($0) })
    }

    private func handleFindStatesBySubstring(substring: String, result: @escaping FlutterResult) {
        guard let geodb = geodb else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "GeoDB not initialized", details: nil))
            return
        }

        let states = geodb.findStatesBySubstring(substr: substring)
        result(states.map { cityResultToMap($0) })
    }

    private func handleFindCitiesBySubstring(substring: String, result: @escaping FlutterResult) {
        guard let geodb = geodb else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "GeoDB not initialized", details: nil))
            return
        }

        let cities = geodb.findCitiesBySubstring(substr: substring)
        result(cities.map { cityResultToMap($0) })
    }

    private func handleFindNearest(lat: Double, lng: Double, count: Int, result: @escaping FlutterResult) {
        guard let geodb = geodb else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "GeoDB not initialized", details: nil))
            return
        }

        let cities = geodb.findNearest(lat: lat, lng: lng, count: UInt32(count))
        result(cities.map { cityResultToMap($0) })
    }

    private func handleFindInRadius(lat: Double, lng: Double, radiusKm: Double, result: @escaping FlutterResult) {
        guard let geodb = geodb else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "GeoDB not initialized", details: nil))
            return
        }

        let cities = geodb.findInRadius(lat: lat, lng: lng, radiusKm: radiusKm)
        result(cities.map { cityResultToMap($0) })
    }

    private func handleSmartSearch(query: String, result: @escaping FlutterResult) {
        guard let geodb = geodb else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "GeoDB not initialized", details: nil))
            return
        }

        let results = geodb.smartSearch(query: query)
        result(results.map { cityResultToMap($0) })
    }

    // MARK: - Helper Methods

    private func cityResultToMap(_ city: CityResult) -> [String: Any] {
        var map: [String: Any] = [
            "name": city.name,
            "state": city.state,
            "country": city.country,
            "iso2": city.iso2,
            "lat": city.lat,
            "lng": city.lng,
            "population": Int(city.population)
        ]

        if let distanceKm = city.distanceKm {
            map["distanceKm"] = distanceKm
        }

        return map
    }
}

File: GeoDB-Apps/geodb_flutter/android/src/main/kotlin/com/example/geodb_flutter/GeodbFlutterPlugin.kt
package com.example.geodb_flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import uniffi.geodb_ffi.GeoDbEngine
import uniffi.geodb_ffi.CityResult
import uniffi.geodb_ffi.DbStatsDto

/** GeodbFlutterPlugin */
class GeodbFlutterPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private var geodb: GeoDbEngine? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "geodb_flutter")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> handleInitialize(result)
            "getStats" -> handleGetStats(result)
            "getCountryCount" -> handleGetCountryCount(result)
            "findCountryByCode" -> {
                val code = call.argument<String>("code")
                if (code == null) {
                    result.error("INVALID_ARGS", "Missing 'code'", null)
                } else {
                    handleFindCountryByCode(code, result)
                }
            }
            "findCountriesBySubstring" -> {
                val substring = call.argument<String>("substring")
                if (substring == null) {
                    result.error("INVALID_ARGS", "Missing 'substring'", null)
                } else {
                    handleFindCountriesBySubstring(substring, result)
                }
            }
            "findStatesBySubstring" -> {
                val substring = call.argument<String>("substring")
                if (substring == null) {
                    result.error("INVALID_ARGS", "Missing 'substring'", null)
                } else {
                    handleFindStatesBySubstring(substring, result)
                }
            }
            "findCitiesBySubstring" -> {
                val substring = call.argument<String>("substring")
                if (substring == null) {
                    result.error("INVALID_ARGS", "Missing 'substring'", null)
                } else {
                    handleFindCitiesBySubstring(substring, result)
                }
            }
            "findNearest" -> {
                val lat = call.argument<Double>("lat")
                val lng = call.argument<Double>("lng")
                val count = call.argument<Int>("count")
                if (lat == null || lng == null || count == null) {
                    result.error("INVALID_ARGS", "Missing lat, lng, or count", null)
                } else {
                    handleFindNearest(lat, lng, count, result)
                }
            }
            "findInRadius" -> {
                val lat = call.argument<Double>("lat")
                val lng = call.argument<Double>("lng")
                val radiusKm = call.argument<Double>("radiusKm")
                if (lat == null || lng == null || radiusKm == null) {
                    result.error("INVALID_ARGS", "Missing lat, lng, or radiusKm", null)
                } else {
                    handleFindInRadius(lat, lng, radiusKm, result)
                }
            }
            "smartSearch" -> {
                val query = call.argument<String>("query")
                if (query == null) {
                    result.error("INVALID_ARGS", "Missing 'query'", null)
                } else {
                    handleSmartSearch(query, result)
                }
            }
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            else -> result.notImplemented()
        }
    }

    // Handler Methods

    private fun handleInitialize(result: Result) {
        try {
            geodb = GeoDbEngine()
            result.success(null)
        } catch (e: Exception) {
            result.error("INIT_ERROR", "Failed to initialize GeoDB", e.message)
        }
    }

    private fun handleGetStats(result: Result) {
        val db = geodb
        if (db == null) {
            result.error("NOT_INITIALIZED", "GeoDB not initialized", null)
            return
        }

        val stats = db.stats()
        result.success(mapOf(
            "countries" to stats.countries.toLong(),
            "states" to stats.states.toLong(),
            "cities" to stats.cities.toLong()
        ))
    }

    private fun handleGetCountryCount(result: Result) {
        val db = geodb
        if (db == null) {
            result.error("NOT_INITIALIZED", "GeoDB not initialized", null)
            return
        }

        result.success(db.countryCount().toInt())
    }

    private fun handleFindCountryByCode(code: String, result: Result) {
        val db = geodb
        if (db == null) {
            result.error("NOT_INITIALIZED", "GeoDB not initialized", null)
            return
        }

        val country = db.findCountryByCode(code)
        if (country != null) {
            result.success(cityResultToMap(country))
        } else {
            result.success(null)
        }
    }

    private fun handleFindCountriesBySubstring(substring: String, result: Result) {
        val db = geodb
        if (db == null) {
            result.error("NOT_INITIALIZED", "GeoDB not initialized", null)
            return
        }

        val countries = db.findCountriesBySubstring(substring)
        result.success(countries.map { cityResultToMap(it) })
    }

    private fun handleFindStatesBySubstring(substring: String, result: Result) {
        val db = geodb
        if (db == null) {
            result.error("NOT_INITIALIZED", "GeoDB not initialized", null)
            return
        }

        val states = db.findStatesBySubstring(substring)
        result.success(states.map { cityResultToMap(it) })
    }

    private fun handleFindCitiesBySubstring(substring: String, result: Result) {
        val db = geodb
        if (db == null) {
            result.error("NOT_INITIALIZED", "GeoDB not initialized", null)
            return
        }

        val cities = db.findCitiesBySubstring(substring)
        result.success(cities.map { cityResultToMap(it) })
    }

    private fun handleFindNearest(lat: Double, lng: Double, count: Int, result: Result) {
        val db = geodb
        if (db == null) {
            result.error("NOT_INITIALIZED", "GeoDB not initialized", null)
            return
        }

        val cities = db.findNearest(lat, lng, count.toUInt())
        result.success(cities.map { cityResultToMap(it) })
    }

    private fun handleFindInRadius(lat: Double, lng: Double, radiusKm: Double, result: Result) {
        val db = geodb
        if (db == null) {
            result.error("NOT_INITIALIZED", "GeoDB not initialized", null)
            return
        }

        val cities = db.findInRadius(lat, lng, radiusKm)
        result.success(cities.map { cityResultToMap(it) })
    }

    private fun handleSmartSearch(query: String, result: Result) {
        val db = geodb
        if (db == null) {
            result.error("NOT_INITIALIZED", "GeoDB not initialized", null)
            return
        }

        val results = db.smartSearch(query)
        result.success(results.map { cityResultToMap(it) })
    }

    // Helper Methods

    private fun cityResultToMap(city: CityResult): Map<String, Any?> {
        return mapOf(
            "name" to city.name,
            "state" to city.state,
            "country" to city.country,
            "iso2" to city.iso2,
            "lat" to city.lat,
            "lng" to city.lng,
            "population" to city.population.toLong(),
            "distanceKm" to city.distanceKm
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        geodb = null
    }
}
